generator/webapp/src/main/java/io/fabric8/maven/generator/webapp/WebAppGenerator.java

```diff
@@ -0,0 +1,190 @@
+ * limitations under the License.
+ */
+package io.fabric8.maven.generator.webapp;
+
+import io.fabric8.maven.core.util.Configs;
+import io.fabric8.maven.core.util.MavenUtil;
+import io.fabric8.maven.docker.config.AssemblyConfiguration;
+import io.fabric8.maven.docker.config.BuildImageConfiguration;
+import io.fabric8.maven.docker.config.ImageConfiguration;
+import io.fabric8.maven.generator.api.MavenGeneratorContext;
+import io.fabric8.maven.generator.api.support.BaseGenerator;
+import io.fabric8.utils.Strings;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author kameshs
+ */
+public class WebAppGenerator extends BaseGenerator {
+
+    private static final String DOCKER_TOMCAT_FROM = "fabric8/tomcat-8";
+    private static final String DOCKER_JETTY_FROM = "fabric8/jetty-9";
+    private static final String DOCKER_WILDFLY_FROM = "jboss/wildfly";
+
+    //TODO need to update the s2i images
+    private static final String VANNILA_S2I_FROM = "fabric8/tomcat-8";
+    private static final String REDHAT_DOCKER_S2I = "jboss-eap-6/eap64-openshift";
+
+    public WebAppGenerator(MavenGeneratorContext context) {
+        super(context, "webapp");
+    }
+
+    private enum Config implements Configs.Key {
+        deploymentDir  {{ d = "/deployments"; }},
+        user           {{ d = "jboss:jboss:jboss"; }},
+        cmd            {{ d = "/opt/tomcat/bin/deploy-and-run.sh"; }},
+        webPort        {{ d = "8080"; }},
+        jolokiaPort    {{ d = "8778"; }};
+
+        public String def() { return d; } protected String d;
+    }
+
+
+    @Override
+    public boolean isApplicable(List<ImageConfiguration> configs) {
+        return shouldAddDefaultImage(configs) &&
+                MavenUtil.hasPlugin(getProject(), "org.apache.maven.plugins:maven-war-plugin");
+    }
+
+    protected Map<String, String> getEnv() {
+        Map<String, String> defaultEnv = new HashMap<>();
+        defaultEnv.put("DEPLOY_DIR", getConfig(Config.deploymentDir));
+        return defaultEnv;
+    }
+
+    @Override
+    public List<ImageConfiguration> customize(List<ImageConfiguration> configs) {
+        ImageConfiguration.Builder imageBuilder = new ImageConfiguration.Builder();
+
+        BuildImageConfiguration.Builder buildBuilder = new BuildImageConfiguration.Builder()
+                .assembly(createAssembly())
+                .from(getFrom())
+                .ports(extractPorts())
+                .cmd(getDockerRunCommand())
+                .env(getEnv());
+        addLatestTagIfSnapshot(buildBuilder);
+        imageBuilder
+                .name(getImageName())
+                .alias(getAlias())
+                .buildConfig(buildBuilder.build());
+        configs.add(imageBuilder.build());
+
+        return configs;
+    }
+
+    protected List<String> extractPorts() {
+        List<String> answer = new ArrayList<>();
+        addPortIfValid(answer, getConfig(Config.webPort));
```

Comment:
@rhuss am not sure am getting this point completely.  can you please give me more details -  typically when you say "WebPort could also be chosen based on ..."